### PR TITLE
Feat/add test cable implementation

### LINKED ISSUE #8: Testing support
Add ability to unit-test channels.

For example, by creating a test cable implementation.

## Example

Having a channel:

```js
import { Channel } from "@anycable/core";

class ChatChannel extends Channel {
  static identifier = "ChatChannel";

  async speak(message) {
    return this.perform("speak", { message });
  }

  async leave() {
    // some custom logic
    return this.disconnect();
  }
}
```

We would like to test it like this (using Jest):

```js
import { ChatChannel } from './channel.js';
import { TestCable } from '@anycable/core/testing';

describe('ChatChannel', () => {
  let channel: ChatChannel
  let cable: TestCable

  beforeEach(() => {
    cable = new TestCable();
    channel = new ChatChannel({ id: '2021' });
    cable.subscribe(channel);
  })

  it('speak perform an action', async () => {
    await channel.speak("hello");
    
    expect(cable.outgoing).toEqual([{message: "hello"}])
  })

  it('disconnects when leave', () => {
    await channel.leave()

    expect(channel.state).toEqual("disconnected");
  })
})
``

### PR DESCRIPTION
**Pull request by TheSeally**
Fix #8 

Added test cable implementation according to example from issue.
If this implementations fits I will also update Readme section in repo.

### CODE REVIEW

File: packages/core/testing/cable/index.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  async subscribe(channel) {
+    channel.connecting(this)
+
+    this.channels[channel.identifier] = channel
+    channel.connected(channel.identifier)
+
+    return Promise.resolve(channel.identifier)
+  }
+
+  async perform(identifier, action, payload = {}) {
```
Comment (palkan) [inline packages/core/testing/cable/index.js:16]: I think, we should check for `identifier` existence here and throw an error if id is absent (like here https://github.com/anycable/anycable-client/blob/0f22fa8cfc3223e2b8d54cbc005a8d650dd76e56/packages/core/cable/index.js#L254)

File: packages/core/testing/cable/index.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    this.outgoing.push(payload)
+
+    return Promise.resolve()
+  }
+
+  async unsubscribe(id) {
+    this.channels[id].close()
```
Comment (palkan) [inline packages/core/testing/cable/index.js:23]: Ditto

**Review by palkan — CHANGES_REQUESTED** [review]
LGTM 👍 
Thanks!

Left a few comments.

File: packages/core/testing/cable/index.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    channel.connecting(this)
+
+    this.channels[channel.identifier] = channel
+    channel.connected(channel.identifier)
+
+    return Promise.resolve(channel.identifier)
+  }
+
+  async perform(identifier, action, payload = {}) {
+    this.outgoing.push(payload)
```
Comment (palkan) [inline packages/core/testing/cable/index.js:17]: > should outgoing array stores action name or only payload (In the example above I see only payload)

Yeah, I think we need actions as well. Maybe, we can use some interface like:

```ts
type PerformedAction = {action: string, payload: object}
this.outgoing.push({action, payload}}
```

Comment (TheSeally) [reply packages/core/testing/cable/index.js]: Added an action for outgoing array. Also added type

File: packages/core/testing/cable/index.js (hunk 1 above)
Comment (TheSeally) [reply packages/core/testing/cable/index.js:16]: Added error throwing for non existing identifier

**Comment by TheSeally** [comment]
Added README section for testing. Also added tests for check non existing identifier (checks that `TestCable` throw error)

File: README.md [hunk 4]
```diff
@@ -276,7 +276,63 @@ export default createCable({
...
+    await channel.perform('ask', { message: 'bye' })
+
+    expect(cable.outgoing).toEqual([
+      { action: 'speak', payload: { message: 'hello' } },
+      { action: 'ask', payload: { message: 'bye' } }
+    ])
+  })
+
+  it('disconnects when leave', async () => {
+    await channel.disconnect()
```
Comment (palkan) [inline README.md:330]:
```suggestion
    await channel.leave()
```

File: README.md [hunk 5]
```diff
@@ -276,7 +276,63 @@ export default createCable({
...
+  let cable: TestCable
+
+  beforeEach(() => {
+    cable = new TestCable()
+    channel = new Channel()
+    cable.subscribe(channel)
+  })
+
+  it('perform an speak action', async () => {
+    await channel.perform('speak', { message: 'hello' })
```
Comment (palkan) [inline README.md:320]: Let's provide an example with the channel's own API:

```suggestion
    await channel.speak('hello')
```

**Review by palkan — APPROVED** [review]
Thanks for you work!

Added a couple of suggestions for the Readme example.

Also, could you please add a change log entry?

File: README.md (hunk 5 above)
Comment (TheSeally) [reply README.md]: Yeah, my bad. Copied code to readme from two sources and forgot to change it

**Comment by TheSeally** [comment]
> Thanks for you work!
> 
> Added a couple of suggestions for the Readme example.
> 
> Also, could you please add a change log entry?

Fix problem in readme. Also added change log entry, but please see if I did it correctly

**Comment by palkan** [comment]
Thanks for your help!

Could you please tell me your name (so I can add to [CultOfMartians](https://cultofmartians.com/tasks/anycable-client-testing.html#task))?

**Comment by TheSeally** [comment]
Thanks for review. My name is Виктор Малий

**Pull request merged**